src/track/globaltrackcache.cpp

```diff
@@ -139,41 +154,111 @@ GlobalTrackCache* volatile GlobalTrackCache::s_pInstance = nullptr;
 
 //static
 void GlobalTrackCache::createInstance(GlobalTrackCacheEvictor* pEvictor) {
-    DEBUG_ASSERT(s_pInstance == nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(!s_pInstance);
     s_pInstance = new GlobalTrackCache(pEvictor);
 }
 
 //static
 void GlobalTrackCache::destroyInstance() {
-    DEBUG_ASSERT(s_pInstance != nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(s_pInstance);
     GlobalTrackCache* pInstance = s_pInstance;
     s_pInstance = nullptr;
     delete pInstance;
 }
 
 //static
 void GlobalTrackCache::deleter(Track* pTrack) {
-    if (s_pInstance != nullptr) {
-        s_pInstance->evict(pTrack);
+    DEBUG_ASSERT(pTrack);
+    // Any access to pTrack is forbidden!! Due to race condition
+    // this pointer might already have been deleted! Only the
+    // cache knows this.
+    if (s_pInstance) {
+        s_pInstance->evictAndDelete(pTrack);
+    } else {
+        // Simply delete unreferenced tracks when the cache is no
+        // longer available. This might but should not happen.
```

Comment:
I think this must not happen, else we have a race condition above if s_pInstance is after the check, but before the evictAndDelete() call. 
Can we make s_pInstance a shared pointer, and destroyInstance() will only drop a strong reference to it.
Or is this a bad idea because we do not know which thread deletes the instance finally. 
An other option would be to make the whole cache static. 
Or use a mutex ..

Reply:
We cannot prevent it, because the cache is not able to control the lifetime of the shared pointers. Once a shared pointer has been handed out by the cache it can be freely copied without the cache knowing.

Making GlobalTrackCache static is not a good idea because of the callback pointer that has to be reset when the receiver.

I don't see any other solution for a singleton implementation that provides substantial improvements over the current implementation.

Reply:
I still think that deleting s_pInstance without protection is dangerous. How about not deleting it but clear() and lock the contained containers with an atomic flag instead? 
Or we can make the mutex static and lock it before accessing s_pInstance. We lock it in the next call anyway.